omen2/object.py

```diff
@@ -188,65 +269,95 @@ def _collect_cascade(self):
         return {}
 
+    @staticmethod
+    def _atomic_apply(obj, changes: Dict[str, Any]):
+        """Atomically apply a dictionary of changes to a python object."""
+        tmpobj = obj.__new__(type(obj))  # new obj, no __init__
+        tmpobj.__dict__ = obj.__dict__.copy()  # copy all attrs to new obj
+        tmpobj._force_apply(changes)
+        obj.__dict__ = tmpobj.__dict__  # swap in new dict (atomic)
+
+    def _force_apply(self, changes):
+        # these attribute sets do not go into the changeset
+        # but setters still get triggered normally
+        with self._suppress_set_changes():
+            for k, v in changes.items():
+                setattr(self, k, v)
+
     def _commit(self):
+        changes = []
+
+        if self.__meta.changes:
+            # apply changes to new obj, side effects of setters, etc
+            changes = self.__meta.changes
+            self._atomic_apply(self, changes)
+            self.__meta.changes = {}
+
         cascade = self._collect_cascade() if self._cascade else {}
 
-        if self._meta.table:
-            self._save()
+        self._save(changes)
 
         for val in self.__dict__.values():
             if isinstance(val, Relation):
-                val.commit(self._meta.table.manager)
+                val.commit(self.__meta.table.manager)
 
         for rel, objs in cascade.items():
             for obj in objs:
                 rel._link_obj(obj)
 
+        self._link_custom_types()
+
     def _remove(self):
         cascade = self._get_related() if self._cascade else {}
 
-        if self._meta.table:
-            table = self._meta.table
+        if self.__meta.table:
+            table = self.__meta.table
             table._remove(self)
 
         for rel, objs in cascade.items():
             for obj in objs:
                 rel.remove(obj)
 
-    def _save(self):
+    def _save(self, keys: Iterable[str]):
         need_id_field = self._need_id()
-        table = self._meta.table
-        if need_id_field or self._meta.new:
+        table = self.__meta.table
+        if need_id_field or self.__meta.new:
             table.insert(self, need_id_field)
-        else:
-            table.update(self)
+        elif keys:
+            table.update(self, keys)
 
-        self._meta.new = False
+        self.__meta.new = False
         self._save_pk()
 
-    def _rollback(self):
-        pk = self._to_pk()
-        db_row = self._meta.table.db_select(pk)
-        if db_row:
-            if len(db_row) > 1:
-                raise OmenMoreThanOneError("more than 1 %s in the db" % self.__class__)
-        self._update(db_row[0])
-
     def __enter__(self):
-        if self._meta and self._meta.table:
-            self._meta.lock.acquire()
-            self._meta.locked = True
+        if self.__meta and self.__meta.table:
+            self.__meta.lock.acquire()
+            self.__meta.locked = True
+            self.__meta.suppress_set_changes = False
```

Comment:
just confirming that this isn't necessary, i.e. `suppress_set_changes` should always already be `False`

Reply:
correct, removed in the chris-mr